### PR TITLE
Semantic segment for YOLOv11

### PR DESCRIPTION
**Pull request by kuazhangxiaoai**
This is a draft PR for semantic segmentation of YOLOv11, just including function and code. The doc about this PR is designing. 
I will commit my doc and pretrained weight as soon as possible.

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
Adds full Semantic Segmentation support to Ultralytics, including datasets, augmentations, dataloaders, model configs, and docs — enabling end-to-end training, validation, prediction, and export for YOLO11 semantic segmentation. 🧩🖼️

### 📊 Key Changes
- New task: Semantic Segmentation
  - New dataset class and pipeline: `SemanticDataset`, `build_semantic_dataset`, and dedicated semseg transforms (mosaic/flip/perspective/format).
  - New augmentations: `SemSegMosaic`, `SemSegRandomFlip`, `SemSegRandomPerspective`, and `SemSegFormat`.
  - New mask utilities: polygon extraction, mask verification/loading, and safer resizing (nearest interpolation).
  - New data validation: `verify_image_and_mask` for color-coded masks.
- New configs and models
  - Model config: `ultralytics/cfg/models/11/yolo11-semseg.yaml` (YOLO11 SemanticSegment head).
  - Global task cfg: `ultralytics/cfg/semseg.yaml` (example settings for semseg).
  - Dataset example: `CityEscapeYOLO.yaml` with class names and RGB color map.
- CLI/engine integration
  - Trainer/Validator updated to recognize `task: "semseg"`.
  - Data build API: `build_semantic_dataset(...)`.
- Documentation
  - New page: Semantic Segmentation guide under docs/tasks (usage, train/val/predict/export).
  - Minor doc fixes in Dataset Explorer examples.
- Typing and minor infra
  - Modernized type hints across codebase (e.g., `list[int]`, `tuple[int, int]`).
  - Dataloader `pin_memory` explicitly set to False in `build_dataloader`.
  - `.gitignore` now ignores `/exp/`.

### 🎯 Purpose & Impact
- End-to-end semseg workflow: Users can now train, validate, predict, and export YOLO11 semantic segmentation models with first-class support. ✔️
- Better data quality and robustness: New mask verification and color-map handling reduce dataset errors; nearest neighbor resizing preserves mask integrity. 🛡️
- Improved augmentations for semseg: Task-specific mosaic/flip/perspective boosts robustness for pixel-level tasks. 🔄
- Clear onboarding: New docs and example configs (including CityEscape YOLO palette) make getting started easier. 📚
- Code quality: Widespread typing updates improve readability and tooling support. 🧹

Example: quick start for Semantic Segmentation
- CLI:
  - Train: `yolo train task=semseg data=CityEscapeYOLO.yaml model=ultralytics/cfg/models/11/yolo11-semseg.yaml imgsz=512`
  - Val: `yolo val task=semseg model=path/to/best.pt data=CityEscapeYOLO.yaml`
  - Predict: `yolo predict task=semseg model=path/to/best.pt source=path/to/images`
- Python:
  ```python
  from ultralytics import YOLO
  model = YOLO("ultralytics/cfg/models/11/yolo11-semseg.yaml")
  model.train(task="semseg", data="CityEscapeYOLO.yaml", imgsz=512, epochs=100)
  ```

Note: The provided `semseg.yaml` contains example local Windows paths; update to your environment before use.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kuazhangxiaoai, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review. An Ultralytics engineer will also assist you here shortly. Please review the following checklist to ensure a smooth integration:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional notes for your PR:
- Thank you for proposing Semantic Segmentation support for YOLO11 — exciting addition! 🎉 Please ensure naming is consistently “YOLO11” across code, configs, docs, and examples.
- The new docs at `docs/en/tasks/semseg.md` currently describe instance segmentation content in several places. Please adjust the language, examples, and media to reflect semantic segmentation to avoid user confusion.
- The dataset example `ultralytics/cfg/datasets/CityEscapeYOLO.yaml` appears to contain an absolute path and seems truncated near the end. Consider switching to a relative/example path and completing all class entries.
- Several example files switch typing from `List/Dict/Tuple` to built-in generics like `list/dict/tuple`. Please confirm compatibility with the project’s supported Python versions and ensure CI passes accordingly.
- If possible, include a small example config (e.g., `ultralytics/cfg/models/11/yolo11-semseg.yaml`) and a minimal training/val snippet in the docs to help users get started quickly.
- Keep changes scoped to the new “semseg” task while avoiding unintended behavior changes in other tasks. This helps reviewers validate and merge faster. ✨

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by kuazhangxiaoai** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Y-T-G** [comment]
@kuazhangxiaoai Some of the commits here do no use your username. You will have to ammend them to use your username for CLA

https://github.com/ultralytics/ultralytics/pull/22253/commits/2321a54a7e7c88c3b5b6039f283852e51e90bc45

Or you can squash all your changes to a single commit

**Comment by kuazhangxiaoai** [comment]
> @kuazhangxiaoai Some of the commits here do no use your username. You will have to ammend them to use your username for CLA
> 
> [2321a54](https://github.com/ultralytics/ultralytics/commit/2321a54a7e7c88c3b5b6039f283852e51e90bc45)
> 
> Or you can squash all your changes to a single commit

I am blocked by CLA asistant, because of invalid account of yanggang. the account whose name is yanggang is an invalida account. How do I deal with it. I'm trying to process the problem. And If failed, I will create a new branch for a new PR again.

**Comment by kuazhangxiaoai** [comment]
I will close this PR, and resubmit a new PR.

**Comment by glenn-jocher** [comment]
Thanks, that works—when you open the new PR, please ensure all commits are authored by your GitHub account to satisfy the CLA; the simplest path is to squash to a single commit with the correct author before pushing:

```bash
git fetch upstream
git checkout -b semseg-new upstream/main
git config user.name "kuazhangxiaoai"
git config user.email "your_verified_email@example.com"
git merge --squash <old-branch>
git commit --author="kuazhangxiaoai <your_verified_email@example.com>" -m "feat: YOLO11 semantic segmentation"
git push -u origin semseg-new
```

Also keep changes minimal to the semseg scope, add a small test and a short docs snippet; details are in the Contributing Guide and CLA steps at [Contributing to Ultralytics](https://docs.ultralytics.com/help/contributing/) and [CLA Signing](https://docs.ultralytics.com/help/CLA/).
